src/ports/postgres/modules/mxgboost/madlib_xgboost.py_in

```diff
@@ -0,0 +1,549 @@
+    if len(importance) == 0:
+        plpy.error("No importance found for any feature")
+    fnames_importances = sorted(
+        [(features[int(k.replace('f',''))], importance[k]) for k in importance],
+        key=itemgetter(1),
+        reverse=True
+    )
+    fnames, f_importance_scores = zip(*fnames_importances)
+    important_features = pd.DataFrame(fnames_importances)
+
+    test_ids = [int(x) for x in test_ids]
+    return (features, pickle.dumps(gbm), params, fnames, f_importance_scores,
+        model_metrics.iloc[:,1].values.tolist(), model_metrics.iloc[:,2].values.tolist(),
+        model_metrics.iloc[:,3].values.tolist(),model_metrics.iloc[:,4].values.tolist(),
+        test_ids)
+
+def xgboost_grid_search(schema_madlib, source_table, id_column, class_label,
+                        list_of_features, list_of_features_to_exclude,
+                        params_str, grid_search_results_tbl, class_weights,
+                        train_set_size, train_set_split_var):
+
+    _assert(table_exists(source_table), "Input table missing")
+    _assert(not table_exists(grid_search_results_tbl), "Output table exists")
+    grid_search_results_tbl_summary = add_postfix(grid_search_results_tbl, '_summary')
+    _assert(not table_exists(grid_search_results_tbl_summary), "Output table exists")
+    params = ast.literal_eval(re.sub("[\\t]","",params_str).strip())
+    params_grid = expand_grid(params)
+    #Save each parameter list in the grid as a row in a distributed table
+    grid_search_params_temp_tbl = unique_string('grid_params')
+    grid_search_params_temp_tbl_df = unique_string('grid_params_df')
+    dist_str = "m4_ifdef(`__POSTGRESQL__', `', `DISTRIBUTED BY (params_index)')"
+    sql = """
+        CREATE TEMP TABLE {grid_search_params_temp_tbl}
+        (
+            params_index int,
+            params text
+        ) {dist_str}
+    """.format(**locals())
+    plpy.execute(sql)
+    sql = """
+        INSERT INTO {grid_search_params_temp_tbl}
+            VALUES ({params_index}, $X${val}$X$);
+    """
+    for indx, val in enumerate(params_grid):
+        plpy.execute(
+            sql.format(
+                val=val,
+                params_index = indx+1, #postgres indices start from 1, so keeping it consistent
+                grid_search_params_temp_tbl=grid_search_params_temp_tbl
+            )
+        )
+
+    if list_of_features.strip() == '*':
+        #Extract feature names from information_schema
+        if list_of_features_to_exclude is None:
+            list_of_features_to_exclude = []
+        discard_features = list_of_features_to_exclude + [class_label, id_column]
+        features = [col for col in get_cols(source_table) if col not in discard_features]
+        list_of_features = ','.join(features)
+    else:
+        features = [f.strip() for f in list_of_features.split(',')]
+
+    grid_size = len(params_grid)
+    sql = """
+        CREATE TEMP TABLE {grid_search_params_temp_tbl_df}
```

Comment:
We should DROP these temp table at the end

Reply:
Since they are temp tables, drop statements are not strictly necessary but I will add them just in case.